### ISSUE TITLE
`no-did-mount-set-state` is triggered by `setState` in ES6 arrow functions

### ISSUE THREAD

**Issue by jbaiter** [yannickcr/eslint-plugin-react#56]
This code triggers the `no-did-mount-set-state` rule:

```
React.createClass({
  //...
  componentDidMount() {
    someClass.onSomeEvent((data) => this.setState({data: data});
  }
  //...
}
```

The error is triggered both when I use an expression and a statement body after the arrow (i.e. with curly braces).
This shouldn't trigger the rule, since the callbacks are obviously not invoked during `componentDidMount` but only at a later time.

**Comment by yannickcr** [comment]
I'm afraid we cannot detect that the callback is not immediately invoked during `componentDidMount`.

In this case I suggest you to rewrite your code not to have setState in `componentDidMount`
or simply ignore the warning for this line:

```js
/*eslint-disable react/no-did-mount-set-state */
someClass.onSomeEvent((data) => this.setState({data: data}));
/*eslint-enable react/no-did-mount-set-state */
```
or
```js
someClass.onSomeEvent((data) => this.setState({data: data})); // eslint-disable-line react/no-did-mount-set-state
```

Sorry for not having a better solution there :|

**Comment by ide** [comment]
Would it be possible to make this option configurable ex: strict (false positives) and loose mode (false negatives)? componentDidMount is a common place to set up event handlers so some kind of lenient loose mode could be helpful.

**Comment by yannickcr** [comment]
Do you have an idea of the logic followed by the rule in strict and loose modes ?

Currently the rule is very simple: if it find a `this.setState(...)` in `componentDidMount` or `componentDidUpdate`it's trigger an error.

**Comment by ide** [comment]
My thinking is that strict mode would use the current behavior, and loose mode would suppress the warning if `this.setState` is called within a function.

**Comment by yannickcr** [comment]
Good idea, it should do the trick.

I will work on this.

**Comment by ide** [comment]
@yannickcr Thanks =)

**Comment by silvenon** [comment]
Awesome, thanks so much for this!
